### PR TITLE
Texture, material loading support for visual meshes

### PR DESCRIPTION
**Pull request by eric-heiden**
Load textures and materials from visual meshes using trimesh.
Adds support for visualizing these textured meshes in the viewers:
image

* **New Features**
  * Full texture & PBR support for meshes: per-mesh texture image/path, roughness, and metallic.
  * Importers extract and apply materials/textures from MJCF, URDF, USD and common mesh files; warn when textures lack UVs.
  * Mesh loader and viewer pipelines accept and propagate texture data; GL renderer supports textures and environment maps with reflection/blending.
  * ModelBuilder.add_usd gains hide_collision_shapes option.
  * Move light source and shadow map with the camera in ViewerGL
  * Improve look of the Allegro hand example where now the visual dex cube is rendered with texture instead of the collision shapes

## Summary by CodeRabbit

* **New Features**
  * Meshes now support image albedo textures, per-mesh roughness/metallic, cached texture hashing, and automatic vertex normal computation; improved multi-mesh import.

* **Viewers & Rendering**
  * Per-mesh albedo textures, environment maps, and PBR reflections added; shading updated for roughness/metallic and environment lighting.

* **Importers / UX**
  * Importers propagate per-mesh material/texture info, warn and disable textures when UVs are missing.

* **Utilities**
  * Public texture utilities for loading, normalization, and stable texture hashing; mesh normal and adjacency helpers added.

* **API**
  * Mesh/mesh-loading and viewer logging now accept texture input and UV-preservation options.

### CODE REVIEW

**Comment by andrewkaufman** [comment]
FYI for USD parsing
- roughness, metallic, normals, opacity (and any other PBR properties) can also have textures, its not just diffuse textures
- mesh uvs (and normals) can be supplied as indexed primvars & the parser currently does not handle this properly
- The newton usd converters both produce assets with these details

I can provide a sample asset to demo if needed.

File: newton/_src/geometry/types.py [hunk 1]
```diff
@@ -163,6 +167,10 @@ def __init__(
             is_solid (bool, optional): If True, mesh is assumed solid for inertia computation (default: True).
             maxhullvert (int, optional): Max vertices for convex hull approximation (default: 64).
             color (Vec3 | None, optional): Optional per-mesh base color (values in [0, 1]).
+            texture_path (str | None, optional): Optional texture image path for the mesh.
```
Comment (mmacklin) [inline newton/_src/geometry/types.py]: Seems there are a lot of code paths handling both the path / bytes inputs. Can it be simplified to one parameter `Union[str, nparray]` that just immediately loads the bytes if it's a path?

File: newton/_src/utils/mesh.py [hunk 2]
```diff
@@ -261,6 +375,514 @@ def create_ellipsoid_mesh(
     return np.array(vertices, dtype=np.float32), np.array(indices, dtype=np.uint32)
 
 
+_texture_url_cache: dict[str, bytes] = {}
```
Comment (mmacklin) [inline newton/_src/utils/mesh.py]: Hmm, seems like kind of bad practice to have a global texture cache sitting somewhere inside a `utils` file - is there somewhere else this could live? e.g.: a static member on `Mesh` maybe?

File: newton/_src/utils/mesh.py [hunk 3]
```diff
@@ -261,6 +375,514 @@ def create_ellipsoid_mesh(
...
+        with urlopen(url, timeout=10) as response:
+            data = response.read()
+        _texture_url_cache[url] = data
+        return data
+    except Exception as exc:
+        warnings.warn(f"Failed to download texture image: {url} ({exc})", stacklevel=2)
+        return None
+
+
+def load_texture_image(texture_path: str | None) -> np.ndarray | None:
```
Comment (mmacklin) [inline newton/_src/utils/mesh.py]: Could maybe just be called `load_texture()` since the `_image` is redundant.

File: newton/_src/utils/mesh.py [hunk 4]
```diff
@@ -261,6 +375,514 @@ def create_ellipsoid_mesh(
...
+                gloss = float(value)
+                if attr_name == "shininess":
+                    gloss = min(max(gloss / 1000.0, 0.0), 1.0)
+                roughness = 1.0 - min(max(gloss, 0.0), 1.0)
+                break
+
+    return roughness, metallic, base_color
+
+
+def load_trimesh_meshes(
```
Comment (mmacklin) [inline newton/_src/utils/mesh.py]: `trimesh` is an implementation detail here right?

In that case I would just call this `load_meshes()`.

Comment (mmacklin) [reply newton/_src/utils/mesh.py]: It may also be worth just calling this `load_meshes_from_file()`, and similarly for `load_texture_from_file()` to make it clear where data is coming from.

File: newton/_src/viewer/gl/opengl.py [hunk 5]
```diff
@@ -23,6 +23,7 @@
 
 from newton.utils import create_sphere_mesh
 
+from ...utils.mesh import compute_vertex_normals_wp
```
Comment (mmacklin) [inline newton/_src/viewer/gl/opengl.py]: Would be nice if we just had `compute_vertex_normals()` and just launched it on CPU when necessary instead of also having the NumPy version.

File: newton/_src/viewer/gl/opengl.py [hunk 6]
```diff
@@ -61,6 +62,58 @@ def check_gl_error():
         print(f"Called from: {''.join(stack[-2:-1])}")
 
 
+def _normalize_texture_image(texture_image: np.ndarray) -> tuple[np.ndarray, int]:
```
Comment (mmacklin) [inline newton/_src/viewer/gl/opengl.py]: Again, I'd remove the `_image` suffix, it's redundant and inconsistent with other methods below like `_upload_texture()`

File: newton/_src/viewer/gl/opengl.py [hunk 7]
```diff
@@ -877,6 +933,20 @@ def __init__(self, title="Newton", screen_width=1920, screen_height=1080, vsync=
         self._last_x, self._last_y = self._screen_width // 2, self._screen_height // 2
         self._key_callbacks = []
         self._key_release_callbacks = []
+
+        self._env_texture = None
+        self._env_intensity = 1.0
+        self._pending_env_map = None
```
Comment (mmacklin) [inline newton/_src/viewer/gl/opengl.py]: I don't think this needs to be called `_pending_env_map` - it can just be None until it's initialized later. Also maybe a more descriptive name like `self._env_path` would be better I think.

File: newton/examples/assets/studio_small_03_1k.jpg
Comment (mmacklin) [inline newton/examples/assets/studio_small_03_1k.jpg:1]: Do we actually have rights to this image?

File: newton/_src/usd/utils.py [hunk 8]
```diff
@@ -761,15 +810,25 @@ def get_mesh(
...
+            # First check if normals is a primvar and use its indices
+            normals_primvar = UsdGeom.PrimvarsAPI(prim).GetPrimvar("normals")
+            normal_indices = None
+            if normals_primvar and normals_primvar.IsIndexed():
+                normal_indices = normals_primvar.GetIndices()
+            # Fall back to direct attribute access for backwards compatibility
+            if normal_indices is None:
+                normals_index_attr = prim.GetAttribute("primvars:normals:indices")
+                if normals_index_attr and normals_index_attr.HasAuthoredValue():
+                    normal_indices = normals_index_attr.Get()
```
Comment (andrewkaufman) [inline newton/_src/usd/utils.py]: This should never return true, if `normals_index_attr.HasAuthoredValue()` is true, then `normals_primvar.IsIndexed()` would also be true & we'd never enter this block.

Comment (andrewkaufman) [reply newton/_src/usd/utils.py]: Oh unless you're protecting for the case that `primvars:normals` doesn't exist but `primvars:normals:indices` does exist? I think that should be considered invalid.

Comment (andrewkaufman) [reply newton/_src/usd/utils.py]: Actually, quite a bit of this normals logic should be inverted. The order of operations should be

- use primvars:normals if they exist
  - check the primvar interpolation to determine vertex, faceVarying, etc
  - use primvars:normals:indices if they exist
- use mesh.GetNormalsAttr() only if the primvar does not exist
  - same for GetNormalsInterpolation(), its only relevant for the non-primvar normals

File: newton/examples/assets/studio_small_03_1k.jpg
Comment (eric-heiden) [reply newton/examples/assets/studio_small_03_1k.jpg:1]: I replaced it now with our own custom env map 🙂

File: newton/_src/usd/utils.py [hunk 9]
```diff
@@ -1040,6 +1040,31 @@ def _find_texture_in_shader(shader: UsdShade.Shader | None, prim: Usd.Prim) -> s
     return None
 
 
+def _get_unconnected_input_value(shader: UsdShade.Shader | None, names: tuple[str, ...]) -> Any | None:
```
Comment (andrewkaufman) [inline newton/_src/usd/utils.py]: I'm guessing this can be simplified with `UsdShade.Input.GetValueProducingAttributes()`

File: newton/_src/utils/mesh.py (hunk 2 above)
Comment (eric-heiden) [reply newton/_src/utils/mesh.py]: I moved this now to a new file `texture.py` but I guess we need to expose this cache to the user somehow?
I can add something like `clear_texture_cache()` in `newton.utils`.

File: newton/_src/usd/utils.py (hunk 9 above)
Comment (eric-heiden) [reply newton/_src/usd/utils.py]: I tried to give this comment to Cursor but it produced even more code I don't understand. If you have some example code how this could be applied here that would be great. Otherwise, I would like to get this merged soon and clean up in a follow-up.

Comment (andrewkaufman) [reply newton/_src/usd/utils.py]: If you call `GetValueProducingAttributes()` then the connections are followed automatically, so in the case from the converters where there is a material interface driving the surface shader values, you can locate the "true value" whether it is connected or not. This applies to the UsdUvTexture.file get above in `_find_texture_in_shader` as well, you can call `GetValueProducingAttributes` there to find out if the file is driven by something else.

https://github.com/NVIDIA-Omniverse/usd-exchange/blob/main/source/core/tests/unittest/testMaterialAlgo.py#L714-L718


Similarly, you could go the other way around from the material interface to the shader inputs that it drives, using `ComputeInterfaceInputConsumersMap`

https://github.com/NVIDIA-Omniverse/usd-exchange/blob/main/source/core/tests/unittest/testMaterialAlgo.py#L251-L264

Comment (andrewkaufman) [reply newton/_src/usd/utils.py]: All that said, if its working now & you want to move on, that can be a cleanup later

**Comment by eric-heiden** [comment]
> Approved but please check loading / mesh processing time doesn't regress.

The benchmarks show the setup times are close to unchanged, the texture loading etc. happens only when the viewer actually has to render textures.

File: newton/_src/usd/utils.py (hunk 9 above)
Comment (eric-heiden) [reply newton/_src/usd/utils.py]: Thanks, Opus 4.6 has figured it out now finally to use `GetValueProducingAttributes()` 🙂

File: newton/_src/usd/utils.py [hunk 10]
```diff
@@ -745,31 +854,63 @@ def get_mesh(
...
     if normals is not None:
         normals = np.array(normals, dtype=np.float64)
-        if mesh.GetNormalsInterpolation() == "faceVarying":
-            # compute vertex normals
-            # try to read primvars:normals:indices (the primvar indexer)
-            normals_index_attr = prim.GetAttribute("primvars:normals:indices")
-            if normals_index_attr:
-                normal_indices = np.array(normals_index_attr.Get(), dtype=np.int64)
-                normals_fv = normals[normal_indices]  # (C,3) expanded
+        if normals_interpolation == "faceVarying":
```
Comment (andrewkaufman) [inline newton/_src/usd/utils.py]:
```suggestion
        if normals_interpolation == UsdGeom.Tokens.faceVarying:
```
